### ISSUE TITLE
CI testing of CUDA

### ISSUE THREAD

**Issue by Infinoid** [tensor-compiler/taco#457]
We have some CI testing, for a limited set of features.  But we need better feedback for when a change breaks something on CUDA.

Github action runners run on [constricted VMs with 2 cores and no GPUs](https://docs.github.com/en/actions/using-github-hosted-runners/about-github-hosted-runners#supported-runners-and-hardware-resources).  Thus, if we want to build and run tests automatically on CUDA, we need to provide the hardware ourselves.

**Comment by Infinoid** [comment]
Some problems with the above approach:

* need to test whether forks can run actions on the self-hosted runner
* if using docker or some other encapsulation/sandboxing, the guest cuda library versions must be carefully installed and match the host driver version
* the github runner can try to update itself, and that update process failed for me
* other users of the GPU can interfere with the test results

**Comment by stephenchouca** [comment]
It definitely makes sense to make the CUDA tests be triggered manually for pull requests. Would there be any issues with having the tests be triggered automatically for pushes to master though, since only developers with write access to the repo can push to master?

**Comment by Infinoid** [comment]
I think it makes sense to run it automatically for pushes to master too.  (Though that would be giving you information after-the-fact, it won't prevent the push.)

**Comment by Infinoid** [comment]
I got a friend to help test the details of forks and unauthorized access to self-hosted runners.  Here's what we found.

* The `Settings` tab in my fork was not visible to him
* The `CUDA build and test` action in my fork was visible to him, but the `Run workflow` UI widget was not visible
* When he forked it, actions were disabled in his fork by default
* When he pushed the green button, he was able to run actions on his fork
* The `CUDA build and test` action was visible to him, with a `Run workflow` UI widget
* Attempts to launch the `CUDA build and test` action died immediately with a "No runners available" error message
* The `Settings` tab in his fork did not list any self-hosted runners

So that all looks sane to me.  If someone wants to run their own self-hosted runner for their own fork, they can.  But only taco collaborators can run code on the official one, and they do so via pushes or manual triggers.
